app/services/reporting/account_reuse_report.rb

```diff
@@ -39,77 +63,255 @@ def account_reuse_emailable_report
+        sp: [{
+          num_sps: 0,
+          num_idv_users: 0,
+          num_all_users: 0,
+        }],
+        agency: [{
+          num_agencies: 0,
+          num_idv_users: 0,
+          num_all_users: 0,
+        }],
+      }
 
-      total_proofed = num_active_profiles
+      sp_reuse_results_idv.each do |result|
+        reuse_results[:sp][result[:num_sps]] = {
+          num_sps: result[:num_sps],
+          num_idv_users: result[:num_idv_users],
+          num_all_users: 0, # Fill it in with 'all' results later
+        }
+      end
+      sp_reuse_results_all.each do |result|
+        if reuse_results[:sp][result[:num_sps]].is_a?(Hash)
+          # Hash exists, so replace the zero placeholder value
+          reuse_results[:sp][result[:num_sps]][:num_all_users] = result[:num_all_users]
+        else
+          reuse_results[:sp][result[:num_sps]] = {
+            num_sps: result[:num_sps],
+            num_idv_users: 0, # Since it didn't exist, fill with zero 'idv' results
+            num_all_users: result[:num_all_users],
+          }
+        end
+      end
+      agency_reuse_results_idv.each do |result|
+        reuse_results[:agency][result[:num_agencies]] = {
+          num_agencies: result[:num_agencies],
+          num_idv_users: result[:num_idv_users],
+          num_all_users: 0, # Fill it in with 'all' results later
+        }
+      end
+      agency_reuse_results_all.each do |result|
+        if reuse_results[:agency][result[:num_agencies]].is_a?(Hash)
+          # Hash exists, so replace the zero placeholder value
+          reuse_results[:agency][result[:num_agencies]][:num_all_users] = result[:num_all_users]
+        else
+          reuse_results[:agency][result[:num_agencies]] = {
+            num_agencies: result[:num_agencies],
+            num_idv_users: 0, # Since it didn't exist, fill with zero 'idv' results
+            num_all_users: result[:num_all_users],
+          }
+        end
+      end
 
-      if !reuse_stats.empty?
-        reuse_stats.each do |result_entry|
-          reuse_total_users += result_entry['num_users']
+      reuse_results.each do |results|
+        if results.length > 1
+          # If there are results, then remove the zero placeholder
+          results[0] = nil
         end
+      end
 
-        if total_proofed > 0
-          reuse_stats.each_with_index do |result_entry, index|
-            reuse_stats[index]['percentage'] = result_entry['num_users'] / total_proofed.to_f
+      sp_reuse_stats = {
+        results: reuse_results[:sp].compact,
+        total_all_users: 0,
+        total_all_percent: 0,
+        total_idv_users: 0,
+        total_idv_percent: 0,
+      }
+      agency_reuse_stats = {
+        results: reuse_results[:agency].compact,
+        total_all_users: 0,
+        total_all_percent: 0,
+        total_idv_users: 0,
+        total_idv_percent: 0,
+      }
+
+      total_proofed = num_active_profiles
+
+      [sp_reuse_stats, agency_reuse_stats].each do |stats|
```

Comment:
another idea would be to make a struct for these instead of a hash? and then we coudl make helper functions and call the `calculate_percentages` on that instead of iterating in a loop?

so lile
```
ReuseStats = Struct.new(:reuslts, :total_all_users, :total_all_percent, :total_idv_users, ...)
```

Reply:
`Struct`s have been implemented!